BJS2-77038 Обменный курс

VladimirEkgardt opened the pull request:
Протестировать можно через testController. Перед мерджем я удалю его

VladimirEkgardt committed:
BJS2-77038 Обменный курс

KolyanGrom commented on src/main/java/faang/school/paymentservice/config/CurrencyRateFetcher.java:
```diff
@@ -0,0 +1,31 @@
+    public void saveCurrency(){
+        Mono<ExchangeRateResponseDto> jsonMono = currencyService.saveExchangeRate();
+
+        jsonMono.subscribe(response->redisService.setValue("USD", response.getRates().get("USD")));
```
Избыточные вызовы:
Сейчас jsonMono.subscribe() вызывается дважды (для USD и EUR), что приводит к двум отдельным HTTP-запросам (если currencyService.saveExchangeRate() выполняет запрос к API). можно обернуть эту логику в:

jsonMono.subscribe(response -> {
redisService.setValue("USD", response.getRates().get("USD"));
redisService.setValue("EUR", response.getRates().get("EUR"));
});

KolyanGrom commented on src/main/java/faang/school/paymentservice/config/CurrencyRateFetcher.java:
```diff
@@ -0,0 +1,31 @@
+//        jsonMono.subscribe(response->redisService.setValue("AUD", response.getRates().get("AUD")));
+        jsonMono.subscribe(response->redisService.setValue("EUR", response.getRates().get("EUR")));
+        log.info("The USD and EUR exchange rates have been updated");
+    }
```
Если currencyService или redisService выбросят исключение, оно будет перехвачено механизмом @Retryable
Поэтому необходимо добавить обработку ошибок внутри .subscribe().

KolyanGrom commented on src/main/java/faang/school/paymentservice/config/RedisConfig.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Bean
+    public RedisConnectionFactory redisConnectionFactory(){
+        return new LettuceConnectionFactory("localhost", 6379);
```
Так не совсем красиво делать - ты зашиваешь порт в код! нужно использовать значения описанные в файле application.yml

добавь туда spring:

redis:
host: localhost
port: 6379

И тогда здесь нужно внедрять через @Value или @ConfigurationProperties.
хардкодить порты не стоит)))

KolyanGrom commented on src/main/java/faang/school/paymentservice/config/WebClientConfig.java:
```diff
@@ -0,0 +1,19 @@
+    private String API_URL;
+
+    @Bean
+    public WebClient webClient(WebClient.Builder builder){
```
Если API недоступен или отвечает медленно, запрос может "висеть" бесконечно.

Попробуй так:
@Bean
public WebClient webClient(WebClient.Builder builder) {
return builder
.baseUrl(API_URL)
.clientConnector(new ReactorClientHttpConnector(
HttpClient.create()
.responseTimeout(Duration.ofSeconds(5)) // Таймаут 5 сек
)
.build();
}

KolyanGrom commented on src/main/java/faang/school/paymentservice/dto/ExchangeRateResponseDto.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.Map;
+
+@Data
+public class ExchangeRateResponseDto {
```
Сделать DTO неизменяемым (@Value).
Добавить валидацию полей. (проверка на null и отрицательные значения напривмер)

VladimirEkgardt replied:
Тут валидация не нужна, так как эта дто - ответ от сервиса. Но неизменяемым сделать- хорошее предложение

KolyanGrom commented on src/main/java/faang/school/paymentservice/service/CurrencyService.java:
```diff
@@ -0,0 +1,49 @@
+        return webClient.get()
+                .uri(getApiUrl())
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError,
```
Проблема: В onStatus для 4xx ошибок используется HttpTimeoutException (это неверно, так как 4xx — это клиентские ошибки, например 400 Bad Request или 401 Unauthorized).

сделай так:

.onStatus(HttpStatusCode::is4xxClientError,
response -> Mono.error(new RuntimeException("Client error: " + response.statusCode())))
.onStatus(HttpStatusCode::is5xxServerError,
response -> Mono.error(new RuntimeException("Server error: " + response.statusCode())))

VladimirEkgardt committed:
1) CurrencyRateFetcher
1.1) Объединил в один subscribe два запроса
1.2) Добавил обработку ошибки
2) RedisConfig
2.1) Вынес переменные host и post в yaml
3) application.yaml
3.1) доавил переменные host и port
4) WebClientConfig
4.1) Добавил таймаут
5) переделал ExchangeRateResponseDto в record
6) CurrencyService
6.1) Поменял ошибки при 400-х и 500-х статусах

SnurnicynKonstantin commented on src/main/resources/application.yaml:
```diff
@@ -3,4 +3,20 @@ server:
+  base_url: "https://openexchangerates.org/"
+  url: "api/latest.json"
+  access_key: "0134aec31a134dec86124871d5fc0cbf"
+#  base: "USD"
```
Эти комменты больше не нужны?

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/config/CurrencyRateFetcher.java:
```diff
@@ -0,0 +1,36 @@
+    public void saveCurrency() {
+        Mono<ExchangeRateResponseDto> jsonMono = currencyService.saveExchangeRate();
+
+//        jsonMono.subscribe(response->redisService.setValue("USD", response.getRates().get("USD")));
```
Эти комменты больше не нужны?

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/CurrencyService.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public Mono<ExchangeRateResponseDto> saveExchangeRate() {
+        return webClient.get()
```
Лучше бы FeignClient использовать. Кода больше бы было. Почему все время избегаешь его?)
Ладно, оставь эту реализацию ради практики.

VladimirEkgardt replied:
Не, в данном случае в тз написано через WebClient)))

VladimirEkgardt committed:
Удалил комментарии

VladimirEkgardt committed:
Merge branch 'kraken-master-stream10' into feature-BJS2-77038-exchange-rate

SnurnicynKonstantin reviewed (changes requested):
Не запускается

<img width="683" alt="image" src="https://github.com/user-attachments/assets/c7170ec9-053e-467a-9289-749e54fd23e8" />

VladimirEkgardt committed:
В application.yaml лишний пробел убрал

VladimirEkgardt committed:
Merge remote-tracking branch 'origin/feature-BJS2-77038-exchange-rate' into feature-BJS2-77038-exchange-rate

SnurnicynKonstantin reviewed (changes requested):
Все та же ошибка
<img width="685" alt="image" src="https://github.com/user-attachments/assets/eabddb6a-61b4-4d1e-933f-744cf91df6c6" />
Два одинаковых ключа в проперти

VladimirEkgardt committed:
Разобрался с переменными из properties, в тестах нашел баг
